core/embed/rust/src/ui/model_tr/layout.rs

```diff
@@ -36,6 +41,52 @@ where
     }
 }
 
+impl<T> ComponentMsgObj for PinPage<T>
+where
+    T: Deref<Target = str>,
+{
+    fn msg_try_into_obj(&self, msg: Self::Msg) -> Result<Obj, Error> {
+        match msg {
+            PinPageMsg::Confirmed => self.pin().try_into(),
+            PinPageMsg::Cancelled => Ok(CANCELLED.as_obj()),
+        }
+    }
+}
+
+impl<T, const N: usize> ComponentMsgObj for ChoicePage<T, N>
+where
+    T: Deref<Target = str>,
+{
+    fn msg_try_into_obj(&self, msg: Self::Msg) -> Result<Obj, Error> {
+        match msg {
+            ChoicePageMsg::Confirmed => self.get_current_choice().try_into(),
+        }
+    }
+}
+
+impl<T> ComponentMsgObj for Bip39Page<T>
+where
+    T: Deref<Target = str>,
+{
+    fn msg_try_into_obj(&self, msg: Self::Msg) -> Result<Obj, Error> {
+        match msg {
+            Bip39PageMsg::Confirmed => self.get_current_choice().try_into(),
+        }
+    }
+}
+
+impl<T> ComponentMsgObj for PassphrasePage<T>
+where
+    T: Deref<Target = str>,
+{
+    fn msg_try_into_obj(&self, msg: Self::Msg) -> Result<Obj, Error> {
+        match msg {
+            PassphrasePageMsg::Confirmed => self.passphrase().try_into(),
+            PassphrasePageMsg::Cancelled => Ok(CANCELLED.as_obj()),
+        }
+    }
+}
```

Comment:
lot of duplication here

how about
```
impl ComponentMsgObj for TextResultComponent {
    fn msg_try_into_obj(&self, msg: Self::Msg) -> Result<Obj, Error> {
        match msg {
            TextResultMsg::Confirmed => self.text_result().try_into(),
            TextResultMsg::Cancelled => Ok(CANCELLED.as_obj()),
        }
    }
}
```
and then implementing `TextResultComponent` as appropriate

Reply:
I find this not so bad because each component is special - some allow `Cancelled`, some send result in `Result(String<X>)` etc.